torch/csrc/jit/codegen/cuda/python_frontend/python_bindings.cpp

```diff
@@ -240,80 +160,127 @@ void initNvFuserPythonBindings(PyObject* module) {
+                    ConstantRecord<torch::jit::fuser::cuda::Int, int64_t>(
+                        {out->index}, val));
+            return out;
           },
           py::return_value_policy::reference)
       .def(
           "define_scalar",
-          [](FusionDefinitionContextManager& self,
-             torch::jit::fuser::cuda::DataType dtype =
-                 torch::jit::fuser::cuda::DataType::Double)
-              -> torch::jit::fuser::cuda::Val* {
-            if (dtype == torch::jit::fuser::cuda::DataType::Double) {
-              return IrBuilder::create<Double>();
-            } else if (
-                dtype == torch::jit::fuser::cuda::DataType::ComplexDouble) {
-              return IrBuilder::create<ComplexDouble>();
-            } else if (dtype == torch::jit::fuser::cuda::DataType::Bool) {
-              return IrBuilder::create<Bool>();
-            } else if (dtype == torch::jit::fuser::cuda::DataType::Int) {
-              return IrBuilder::create<Int>();
-            } else {
-              TORCH_CHECK(false, "Dtype is not supported:", dtype);
-            }
+          [](nvfuser::FusionDefinition& self,
+             NvfDataType dtype = torch::jit::fuser::cuda::DataType::Double)
+              -> nvfuser::Scalar* {
+            nvfuser::Scalar* out =
+                new nvfuser::Scalar(self.recording_state.size());
+            self.recording_state.emplace_back(out);
+            self.recording.emplace_back(
+                new nvfuser::ScalarRecord({out->index}, dtype));
+            return out;
           },
           py::arg("dtype") = torch::jit::fuser::cuda::DataType::Double,
           py::return_value_policy::reference);
 
-  py::class_<FusionDefinitionContextManager::Ops> nvf_ops(fusion_def, "Ops");
+  py::class_<nvfuser::FusionDefinition::Operators> nvf_ops(
+      fusion_def, "Operators");
+  nvf_ops.def(py::init<nvfuser::FusionDefinition*>());
 
   // ******************** INSERT OP BINDINGS BELOW HERE ********************
 
-#define NVFUSER_PYTHON_BINDING_UNARY_OP(op_str, op_name)                 \
-  nvf_ops.def_static(                                                    \
-      op_str,                                                            \
-      py::overload_cast<TensorView*>(&torch::jit::fuser::cuda::op_name), \
-      py::return_value_policy::reference);                               \
-  nvf_ops.def_static(                                                    \
-      op_str,                                                            \
-      py::overload_cast<torch::jit::fuser::cuda::Val*>(                  \
-          &torch::jit::fuser::cuda::op_name),                            \
+#define NVFUSER_PYTHON_BINDING_UNARY_OP(op_str, op_name)              \
+  nvf_ops.def(                                                        \
+      op_str,                                                         \
+      [](nvfuser::FusionDefinition::Operators& self,                  \
```

Comment:
nitpick.. see many boiler plates here. I'm wondering if changing these to a template would make it easier to maintain...

Reply:
I went through a changed up the binding lambda functions to remove 2 lines per definition that were pure boiler plate code.  Here is an example:

Old:

```
#define NVFUSER_PYTHON_BINDING_UNARY_OP(op_str, op_name)              \
  nvf_ops.def(                                                        \
      op_str,                                                         \
      [](nvfuser::FusionDefinition::Operators& self,                  \
         nvfuser::Tensor* input) -> nvfuser::Tensor* {                \
        nvfuser::Tensor* output = new nvfuser::Tensor(                \
            self.fusion_definition->recording_state.size());          \
        self.fusion_definition->recording_state.emplace_back(output); \
        self.fusion_definition->recording.emplace_back(               \
            new nvfuser::OpRecord<NvfTensorView*, NvfTensorView*>(    \
                {input->index},                                       \
                {output->index},                                      \
                static_cast<NvfTensorView* (*)(NvfTensorView*)>(      \
                    torch::jit::fuser::cuda::op_name)));              \
        return output;                                                \
      },                                                              \
      py::return_value_policy::reference);                            \
```

New:

```
#define NVFUSER_PYTHON_BINDING_UNARY_OP(op_str, op_name)                  \
  nvf_ops.def(                                                            \
      op_str,                                                             \
      [](nvfuser::FusionDefinition::Operators& self,                      \
         nvfuser::Tensor* input) -> nvfuser::Tensor* {                    \
        nvfuser::Tensor* output = self.fusion_definition->defineTensor(); \
        self.fusion_definition->defineRecord(                             \
            new nvfuser::OpRecord<NvfTensorView*, NvfTensorView*>(        \
                {input->index},                                           \
                {output->index},                                          \
                static_cast<NvfTensorView* (*)(NvfTensorView*)>(          \
                    torch::jit::fuser::cuda::op_name)));                  \
        return output;                                                    \
      },                                                                  \
      py::return_value_policy::reference);                                \
```

Reply:
I'm actually wondering if a generic lambda would allow us to merge the variants between scalar and tensor here. But we can leave code cleaning to a separate PR.

Reply:
Are you suggesting a variadic templatized lambda?